src/langcheck/metrics/custom_text_quality.py

```diff
@@ -0,0 +1,104 @@
+
+
+def custom_evaluator(generated_outputs: list[str] | str | None,
+                     prompts: list[str] | str | None,
+                     sources: list[str] | str | None,
+                     reference_outputs: list[str] | str | None,
+                     eval_model: EvalClient, metric_name: str,
+                     score_map: dict[str, float], template_path: str,
+                     language: str) -> MetricValue[float | None]:
+    '''Calculates the scores of a custom evaluator. The EvalClient will first
+    assess the provided inputs using the prompt template, and then convert those
+    assessments into scores using the score map.
+
+    Args:
+        generated_outputs: The model generated output(s)
+        prompts: The prompts used to generate the output(s)
+        sources: The source(s) of the generated output(s)
+        reference_outputs: The reference output(s)
+        eval_model: The EvalClient instance used for the evaluation
+        metric_name: The name of the metric
+        score_map: A dictionary mapping the evaluator's assessments to scores
+        template_path: The path to the prompt template file
+        language: The language that the evaluator will use ('en', 'ja', or 'de')
+
+    Returns:
+        A MetricValue object
+    '''
+    generated_outputs, prompts, reference_outputs, sources = validate_parameters_custom_evaluator(  # NOQA: E501
+        generated_outputs, prompts, reference_outputs, sources)
+    # Find the length of the first non-None list (they are guaranteed to all be
+    # the same length)
+    num_examples = next(
+        (len(lst)
+         for lst in [generated_outputs, prompts, reference_outputs, sources]
+         if lst is not None), 0)
+
+    if language not in ['en', 'ja', 'de']:
+        raise ValueError(f'Unsupported language: {language}')
+
+    assert Path(template_path).exists(
+    ), f'Prompt template file {template_path} does not exist.'
+    prompt_template_source = Path(template_path).read_text()
+    prompt_template = Template(prompt_template_source)
+
+    # Validate the expected parameters in the prompt template
+    env = Environment()
+    expected_params = meta.find_undeclared_variables(
+        env.parse(prompt_template_source))
+    allowed_params = ['gen_output', 'user_query', 'src', 'ref_output']
+    assert all(param in allowed_params for param in expected_params), \
+        f'The prompt template contains invalid parameters. The allowed parameters are {allowed_params} but the prompt template expects the parameters {expected_params}'  # NOQA: E501
+    expected_param_to_arg = {
+        'gen_output': generated_outputs,
+        'user_query': prompts,
+        'src': sources,
+        'ref_output': reference_outputs,
+    }
+    for param in expected_params:
+        assert expected_param_to_arg[param] is not None, \
+            f'The prompt template expects the parameter "{param}" but it is not provided.'  # NOQA: E501
```

Comment:
You might want to add `{% if src is none... %}` condition as we do for pairwise comparison, maybe it is better to remove the validation?

Reply:
That's a fair point, but I still think it's worth validating this, since the behavior can be surprising otherwise. E.g. if the user has `{% if src is none... %}` but then passes in `sources=None`, the individual `src` vars actually become empty strings and not `None`.

So if the user has logic such as `{% if src is none... %}` in their prompt template, IMO it's better for them to pass in a list of Nones as `sources`, WDYT?

Reply:
Makes sense!